### DISCUSSION TITLE
Question: how do users normally compare their own structs?

### DISCUSSION THREAD

**Discussion by rgwood** [Byron/treediff-rs#10] [Q&A] [answered]
Hi! This looks like a really cool+useful crate. I've got a question that is not really an issue but this felt like the best place to ask it.

I've got a handful of structs that I would like to diff. They're pretty simple (strings, ints, booleans) and they all implement Serialize+Deserialize from `serde`. Implementing `Value` manually on each structs feels like a hassle, so instead I'm serializing to JSON before diffing:

```rust
let item1_json: serde_json::Value = item1.serialize(serde_json::value::Serializer)?;
let item2_json: serde_json::Value = item2.serialize(serde_json::value::Serializer)?;
let mut recorder = Recorder::default();
treediff::diff(&item1_json, &item2_json, &mut recorder);
```

That works, but I'm still quite new to treediff and I'm wondering if that's the best approach. **Is there a simpler approach that I'm missing?** It would be really cool if treediff automatically worked on anything that implements Serialize+Deserialize (but I'm probably missing a good reason why it can't).

**Comment by Byron** [comment, marked as answer]
Thanks for making me aware the discussions weren't activated here :) - and now that this is fixed, let me share my thoughts on this.

First of all, I agree that nowadays it could and maybe should support `serde` out of the box to remove the extra conversion step.

However, I also feel a bit uneasy about this crate as it was from a time where Rust was new to me and where I had a python mindset. This crate is essentially a rewrite of something I had and enjoyed in the python world. This crate is now at a point where it probably works well enough, but there is at least one unresolved issues which seems to hint at a bug in the implementation. And I can't fix it I am not even using this crate anymore.

From that point of view it's probably better to archive it, but then people wouldn't be able to contribute fixes either, so I don't know if there is more harm in letting people pick this crate up compared to avoiding it, but preventing maintenance.

Maybe, just maybe, it's time for something new? Could there be a diff-crate whose sole feature is to calculate changes between two structures implementing `serde::Serialize`? There seems to be a `serde-diff` crate which can diff two instances of the same type, so that would need generalization.

----

In short, this crate has bugs, isn't the most well-tested, and I don't put any time into furthering its development, and it's definitely wasteful with resources. It could use a modern rewrite that doesn't have any of these issues, and I don't recommend proliferating this crate by adopting it. There really should be alternatives, or somebody taking over the development here to fix issues.

**Reply by rgwood** [reply]
Thanks for the quick answer! Good to have that context, I'll look for another solution.

[It looks like `serde-diff` is currently unmaintained](https://github.com/amethyst/serde-diff/issues/42); it seems like this is a problem without a great solution in the crate ecosystem right now.
